ouroboros-consensus/src/Data/FingerTree/TopMeasured/Strict.hs

```diff
@@ -0,0 +1,164 @@
+  , measureInternal
+    -- * API
+  , empty
+  , fmap'
+  , fmap''
+  , fromList
+  , split
+  , (|>)
+  ) where
+
+import           Data.Foldable
+import           Data.Group
+import           GHC.Generics (Generic)
+
+import           NoThunks.Class (NoThunks (..), noThunksInValues)
+
+import           Data.FingerTree.Strict (Measured)
+import qualified Data.FingerTree.Strict as FT
+
+-- TODO(jdral): Should we force strictness anywhere? Or more generally, where
+-- should we guide evaluation?
+
+{-------------------------------------------------------------------------------
+  Strict finger trees with top-level measures
+-------------------------------------------------------------------------------}
+
+-- | A @StrictFingerTree@ with elements of type @a@, an internal measure
+-- of type @vi@, and a top-level measure of type @vt@.
+data StrictFingerTree vt vi a = StrictFingerTree  {
+    tm       :: vt
+  , elements :: !(FT.StrictFingerTree vi a)
+  }
+  deriving (Show, Eq, Ord, Generic)
+
+instance Foldable (StrictFingerTree vt vi) where
+  foldMap f = foldMap f . elements
+
+instance NoThunks a => NoThunks (StrictFingerTree vt vi a) where
+  showTypeOf _ = "Alt"
+  wNoThunks ctxt = noThunksInValues ctxt . toList
+
+instance (Semigroup vt, Measured vi a)
+      => Semigroup (StrictFingerTree vt vi a) where
+  StrictFingerTree tm1 xs1 <> StrictFingerTree tm2 xs2 =
+    StrictFingerTree (tm1 <> tm2) (xs1 <> xs2)
+
+instance (Monoid vt, Measured vi a) => Monoid (StrictFingerTree vt vi a) where
+  mempty = StrictFingerTree mempty mempty
+
+{-------------------------------------------------------------------------------
+  Measuring
+-------------------------------------------------------------------------------}
+
+-- | Internal measures are used by the internal @'StrictFingerTree'@.
+type InternalMeasured v a = Measured v a
+
+measureInternal :: Measured v a => a -> v
+measureInternal = FT.measure
+
+-- | All values of type @Alt@ are internal-measured.
+instance Measured vi a => Measured vi (StrictFingerTree vt vi a) where
+  measure = FT.measure . elements
+
+-- | Re-iteration of @'Measured'@, but for top-level measures.
+--
+-- This re-iteration is necessary because we want to allow the top-level measure
+-- to be distinct from the internal measure. For example, we can not create both
+-- of these instances:
+-- > @instance Measured T  a where -- ...@
+-- > @instance Measured T' a where -- ...@
+-- Furthermore, we want the top-level measure to be a @'Group'@ instead of a
+-- @'Monoid'@.
+class Group v => TopMeasured v a | a -> v where
+  measureTop :: a -> v
+
+-- | All values of type @Alt@ are top-measured.
+instance TopMeasured vt a => TopMeasured vt (StrictFingerTree vt vi a) where
+  measureTop = tm
+
+-- | Wrapper for when we need both @TopMeasured@ and @InternalMeasured@
+type SuperMeasured vt vi a = (TopMeasured vt a, InternalMeasured vi a)
+
+{-------------------------------------------------------------------------------
+  API
```

Comment:
This is a minimal API based on what `consensus` requires. Should we extend this API or is it out of the scope of this PR? I can imagine that we would want to extend this API only when we decide whether to integrate the new anti-diff approach in the `consensus` code. This integration would also be a separate issue/PR.